pkg/reconciliation/generic.go

```diff
@@ -18,7 +18,8 @@ type Reconcileable[T any] interface {
 	*T
 }
 
-// Try with U, a type of any whose POINTER still fulfils Reoncilable...
+// ReconcileObject ensures that desiredObject exists in the given state, either by creating it, or updating it if it
+// already exists.
 func ReconcileObject[U any, T Reconcileable[U]](ctx context.Context, kClient client.Client, requeueDelay time.Duration, desiredObject U) result.ReconcileResult {
```

Comment:
[thought] This is off-topic, but I don't understand why this function schedules a requeue after a successful creation.

Reply:
Because that's the pattern followed throughout the codebase. Object creation is asynchronous, acceptance by the API server does not necessarily mean that something won't go wrong downstream of that. As a result, we check again for the success of the creation operation before continuing, which indeed does mean a requeue.

There is stuff to do with optimistic locking and the way the caches work on the controller which make this necessary - try removing it if you're curious as to what breaks.

Reply:
> Object creation is asynchronous, acceptance by the API server does not necessarily mean that something won't go wrong downstream of that.

Is that documented anywhere? The `client` package docs do not go into many details.

Reply:
That's just how asynchronous HTTP APIs work. But I think the more critical thing here is that (from memory) due to some of the caching behaviour of these clients, the object will not appear if you try to go do a Get in some circumstances. I encourage you to try to remove the requeues and see if things break - I believe they do.

Reply:
I don't know, [this example in the controller-runtime codebase](https://github.com/kubernetes-sigs/controller-runtime/blob/main/examples/crd/main.go#L90) doesn't requeue. They reference `pod.CreationTimestamp`, a field that is set by the server during the creation operation, immediately after a successful `Create()` call.

Reply:
I'm not sure, my understanding was that at least caching meant that requeues were necessary. But I'm open to being wrong here :)

Reply:
No, if write is success, it really is written. Writing is not an asynchronous operation in Kubernetes, it's actually strictly consistent one (using Raft).

Reading is asynchronous however as we use a cache.

Reply:
OK, based on this I've modified ReconcileObject to remove those extra requeues. It's in a separate commit if we decide to revisit it later.

Reply:
> OK, based on this I've modified ReconcileObject to remove those extra requeues. It's in a separate commit if we decide to revisit it later.

That's incorrect @olim7t - based on what Micke has just said above `Reading is asynchronous however as we use a cache.` That means that (while I'm wrong on the consistency level of writes, I'm right on reads, and some of our logic will probably fail due to failure to refresh the cache.

Reply:
By "extra" requeues I meant the one after a successful Create, and the one after a successful Update. I pushed the change so you can check the diff.

Reply:
> That means that (while I'm wrong on the consistency level of writes), I'm right on reads, and some of our logic will probably fail due to failure to refresh the cache.

This depends on what is read after the object was written. If it's something that the Kubernetes API returns, then it's consistent right away. For example, Olivier's example of using creationTimestamp, probably resourceVersion and uid etc are directly accessible in the object you just posted.

However, for many use cases, what happens is that there's other controllers doing some extra processing and then modifying the object on the Kubernetes side (as a lot of Kubernetes processing itself is async with controllers). Can't access or Update .Status for example for some objects, like Pods. Or in our runtimes, if k8ssandra-operator creates CassandraDatacenter, it probably can't Update the Status field of CassandraDatacenter even instantly after, because cass-operator has modified it and the resourceVersions no longer match.

The other thing that should be understood is that Read after Write is cached, but of course Write itself updates the object that was sent. So doing:

obj = Create(obj)
a := obj.ObjectMeta.uid
// and if nothing else is touching our object, we can patch it afterwards also.

Is valid. However, doing:

obj = Create(obj)
obj = Get(obj)
a := obj.ObjectMeta.uid

Is not valid. The Get() call is coming from cache and would probably not return an object at all. Using Reader interface from manager will allow doing this as it isn't cached, but we don't really use that interface a lot (I think only ClientConfigs does that).

Reply:
What we're saying here is that a requeue may or may not be needed, depending on the case.
If the requeue itself is harmless, just adding a little delay, but is protecting the code from potential bugs we might as well keep it, don't we?
@olim7t, for the sake of moving this PR forward and keeping its scope to (mostly) the original one, I'd suggest to rollback this change and have a separate discussion for it.

Reply:
> What we're saying here is that a requeue may or may not be needed, depending on the case.

This thread has turned into a generic conversation about requeues, but we are in fact looking at a very specific case. I've only made two changes related to requeues in this PR:
- ReconcileObject tries to create the object and succeeds
- ReconcileObject tries to update the object and succeeds

(it's all in [Don't requeue successful writes in ReconcileObject](https://github.com/k8ssandra/k8ssandra-operator/pull/1463/commits/5bbd6af0294d3b8a2072ab1c494773771019c433) if anyone wants to look at the diff)

If we accept the assertion that writes are synchronous then there's absolutely no need to requeue. ReconcileObject wanted to apply changes, those changes are applied, we're done.

Other controllers and cached reads are valid concerns that we encounter in other places, but not here.